File: src/main/java/com/shimmermare/inviteroles/InviteRoles.java
/*
 * MIT License
 *
 * Copyright (c) 2019 Shimmermare
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.shimmermare.inviteroles;

import com.mojang.brigadier.CommandDispatcher;
import joptsimple.OptionParser;
import joptsimple.OptionSet;
import joptsimple.OptionSpec;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.JDABuilder;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.User;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.security.auth.login.LoginException;
import java.io.IOException;
import java.io.InputStream;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class InviteRoles
{
    private static final Logger LOGGER = LoggerFactory.getLogger(InviteRoles.class);

    private static final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();

    static
    {
        executorService.execute(() -> Thread.currentThread().setName("DBUpdater"));
    }

    private final String token;
    private final String dbPath;

    private String version;

    private JDA jda;
    private DatabaseService databaseService;
    private CommandDispatcher<CommandSource> commandDispatcher;
    private Map<Long, ServerInstance> joinedServers = new ConcurrentHashMap<>();

    public InviteRoles(String token, String dbPath)
    {
        this.token = token;
        this.dbPath = dbPath;
    }

    private void readProperties()
    {
        LOGGER.debug("Reading application properties from resource");
        Properties properties = new Properties();
        try (InputStream is = InviteRoles.class.getResourceAsStream("/.properties"))
        {
            properties.load(is);
        }
        catch (Exception e)
        {
            LOGGER.error("Unable to load application properties", e);
        }

        this.version = properties.getProperty("version", "unknown");
    }

    private void run()
    {
        readProperties();

        LOGGER.info("Starting InviteRoles Discord Bot version {}. I'm alive!", this.version);

        commandDispatcher = new CommandDispatcher<>();
        Command.register(commandDispatcher);

        JDABuilder jdaBuilder = new JDABuilder(token);

        try
        {
            jda = jdaBuilder
                    .addEventListeners(new EventListener(this))
                    .build();
        }
        catch (LoginException e)
        {
            LOGGER.error("Unable to login to Discord", e);
            System.exit(-1);
        }

        try
        {
            jda.awaitReady();
        }
        catch (InterruptedException e)
        {
            LOGGER.error("Failed to await JDA ready", e);
            System.exit(-1);
        }

        User selfUser = jda.getSelfUser();
        LOGGER.info("Successfully logged in as {} ({}).", selfUser.getName(), selfUser.getIdLong());

        databaseService = DatabaseService.connect(dbPath);
        if (databaseService == null)
        {
            System.exit(-1);
        }
        databaseService.checkTables();

        checkGuilds();

        executorService.scheduleAtFixedRate(() ->
        {
            for (ServerInstance instance : joinedServers.values())
            {
                if (instance.isUpdated())
                {
                    long server = instance.getServer().getIdLong();
                    databaseService.setServerSettings(server, instance.getServerSettings());
                    instance.setUpdatedFlag(false);
                    LOGGER.debug("Database info of server {} updated by scheduler.", server);
                }
            }
        }, 5L, 5L, TimeUnit.MINUTES);
    }

    private void checkGuilds()
    {
        List<Guild> servers = jda.getGuilds();
        int totalMembers = 0;
        for (Guild server : servers)
        {
            totalMembers += server.getMembers().size();

            ServerSettings settings = databaseService.getServerSettings(server.getIdLong());
            if (settings == null)
            {
                newServerInstance(server);
            }
            else
            {
                ServerInstance instance = newServerInstance(server, settings);
                instance.checkInviteRoles();
            }
        }
        LOGGER.info("Currently joined {} servers with a total of {} members.", servers.size(), totalMembers);
    }

    public void forceUpdateDatabase()
    {
        for (ServerInstance instance : joinedServers.values())
        {
            long server = instance.getServer().getIdLong();
            databaseService.setServerSettings(server, instance.getServerSettings());
            instance.setUpdatedFlag(false);
        }
    }

    public void shutdown()
    {
        jda.shutdown();

        forceUpdateDatabase();
        databaseService.close();

        System.exit(0);
    }

    public Map<Long, ServerInstance> getJoinedServers()
    {
        return Collections.unmodifiableMap(joinedServers);
    }

    public ServerInstance getServerInstance(long server)
    {
        return joinedServers.get(server);
    }

    public ServerInstance newServerInstance(Guild server)
    {
        return newServerInstance(server, new ServerSettings());
    }

    public ServerInstance newServerInstance(Guild server, ServerSettings settings)
    {
        ServerInstance instance = new ServerInstance(this, server, settings);
        joinedServers.put(server.getIdLong(), instance);
        return instance;
    }

    public ServerInstance removeServerInstance(long server)
    {
        ServerInstance instance = joinedServers.remove(server);
        if (instance != null)
        {
            databaseService.deleteServerSettings(server);
        }
        return instance;
    }

    public CommandDispatcher<CommandSource> getCommandDispatcher()
    {
        return commandDispatcher;
    }

    public String getDbPath()
    {
        return dbPath;
    }

    public String getVersion()
    {
        return version;
    }

    public JDA getJda()
    {
        return jda;
    }

    public static void main(String[] args)
    {
        OptionParser optionParser = new OptionParser();
        OptionSpec<String> tokenSpec = optionParser.accepts("token").withRequiredArg().ofType(String.class).required();
        OptionSpec<String> dbPathSpec = optionParser.accepts("db").withRequiredArg().ofType(String.class).required();

        OptionSet optionSet = optionParser.parse(args);
        String token = tokenSpec.value(optionSet);
        String dbPath = dbPathSpec.value(optionSet);

        InviteRoles inviteRoles = new InviteRoles(token, dbPath);
        inviteRoles.run();

        try
        {
            System.in.read();
            LOGGER.info("Application terminated from console. Goodbye!");
            inviteRoles.shutdown();
            System.exit(0);
        }
        catch (IOException e)
        {
            LOGGER.error("Unable to read from console", e);
        }
    }
}

File: src/main/java/com/shimmermare/inviteroles/InviteRoles.kt
package com.shimmermare.inviteroles

import com.mojang.brigadier.CommandDispatcher
import joptsimple.OptionParser
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.JDABuilder
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.utils.cache.CacheFlag
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.awt.Color
import java.io.IOException
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import javax.security.auth.login.LoginException
import kotlin.collections.HashSet
import kotlin.system.exitProcess

/**
 * Main bot class. Not valid until run() is called.
 */
class InviteRoles(private val token: String, dbPath: String) {
    val settingsRepository = SettingsRepository(dbPath)
    val invitesRepository = InviteRepository(dbPath)

    // Accessible after run
    lateinit var properties: Properties
        private set
    private lateinit var jda: JDA
    lateinit var commandDispatcher: CommandDispatcher<CommandSource>
        private set

    private val guilds = ConcurrentHashMap<Long, BotGuild>()

    /**
     * Run bot with this action sequence:
     * 1. Load app properties from classpath.
     * 2. Init repositories.
     * 3. Create JDA and setup Discord connection.
     * 4. Fetch settings and invites of joined guilds from repositories.
     * 5. Clear removed invites and roles.
     * 6. Init command dispatcher.
     * 7. Register event listener.
     * 8. Wait for console commands.
     *
     * This function will NEVER return. The only way to exit is to enter "stop" console command
     * or to call [stop].
     */
    fun run() {
        loadProperties()

        settingsRepository.initTable()
        invitesRepository.initTable()

        connectToDiscord()
        fetchJoinedGuilds()
        clearRemovedInvitesAndRoles()

        initCommandDispatcher()
        jda.addEventListener(EventListener(this))

        log.info("Starting InviteRoles Discord Bot v${properties.getProperty("version")}.")

        // Wait for console commands in main thread; bot itself runs in separate threads
        // Call bot.stop() to exit program
        while (true) {
            when (val line: String? = readLine()) {
                "stop" -> {
                    log.info("Received stop console command")
                    stop()
                }
                else -> {
                    println("Unknown console command. Enter 'stop' to stop the bot.")
                    log.info("Received unknown console command: $line")
                }
            }
        }
    }

    /**
     * Load properties from .properties on classpath
     */
    @Throws(IOException::class)
    private fun loadProperties() {
        properties = Properties()

        try {
            val input = InviteRoles::class.java.getResourceAsStream("/.properties")
            input.use(properties::load)
            log.debug("Properties loaded from file")
        } catch (e: IOException) {
            log.error("Failed to load application properties", e)
            throw e
        }
    }

    /**
     * Build JDA and setup discord connection
     */
    @Throws(LoginException::class)
    private fun connectToDiscord() {
        try {
            jda = JDABuilder
                .create(
                    token,
                    GatewayIntent.GUILD_MEMBERS, // FYI: Discord will limit bot to only 100 guilds because of this line
                    GatewayIntent.GUILD_INVITES,
                    GatewayIntent.GUILD_MESSAGES,
                    GatewayIntent.DIRECT_MESSAGES
                )
                .disableCache(CacheFlag.ACTIVITY)
                .disableCache(CacheFlag.CLIENT_STATUS)
                .disableCache(CacheFlag.VOICE_STATE)
                .disableCache(CacheFlag.EMOTE)
                .build()
            log.info("Logging in Discord...")
        } catch (e: LoginException) {
            log.error("Failed to login to Discord!", e)
            throw e
        }

        jda.awaitReady()
        val selfUser: User = jda.selfUser
        log.info("Logged in as ${selfUser.name} (${selfUser.id}).")
    }

    /**
     * Find settings and invites of the joined guilds in the repository
     */
    private fun fetchJoinedGuilds() {
        jda.guilds
            .also { guilds -> log.info("Joined ${guilds.size} guilds.") }
            .forEach { guild ->
                if (!loadGuild(guild)) {
                    log.warn(
                        "Guild {} ({}) doesn't have settings in repository, possible DB corruption",
                        guild.name, guild.id
                    )
                }
            }
    }

    /**
     * Remove invites that have been removed on server or they don't have any existing role
     *
     * This is useful to catch stuff that was deleted when event listener wasn't active
     */
    private fun clearRemovedInvitesAndRoles() {
        getGuilds().forEach { guild ->
            val existingInvites = guild.guild.retrieveInvites().complete().map { it.code } // TODO coroutine candidate
            guild.getInvites().forEach { invite ->
                if (!existingInvites.contains(invite.code)) {
                    log.debug("Invite {} doesn't exist in guild {} and will be removed", invite.code, guild.id)
                    guild.removeInvite(invite)
                }
                if (guild.guild.getRoleById(invite.roleId) == null) {
                    guild.removeInvite(invite)
                    log.debug(
                        "Role {} of invite {} doesn't exist in guild {} and invite will be removed",
                        invite.roleId,
                        invite.code,
                        guild.id
                    )
                }
            }
        }
    }

    /**
     * Create command dispatcher and register commands
     */
    private fun initCommandDispatcher() {
        commandDispatcher = CommandDispatcher<CommandSource>()
        Commands.register(commandDispatcher)
    }

    /**
     * Returns true if bot was used in this guild already
     */
    @Synchronized
    fun loadGuild(guild: Guild): Boolean {
        val settingsFromRepo = settingsRepository.find(guild)
        val invites = invitesRepository.findAllOfGuild(guild)
        addGuild(guild, settingsFromRepo, invites) // Pass settings if null - they'll be defaulted and saved
        return settingsFromRepo != null
    }

    @Synchronized
    fun getGuild(guild: Guild): BotGuild? {
        return guilds[guild.idLong]
    }

    @Synchronized
    fun getGuild(id: Long): BotGuild? {
        return guilds[id]
    }

    @Synchronized
    @Throws(IllegalStateException::class)
    fun getGuildOrThrow(guild: Guild): BotGuild {
        return guilds[guild.idLong] ?: throw IllegalStateException("Guild ${guild.id} is not wrapped by the bot")
    }

    /**
     * Track new guild.
     *
     * @param settings if null, default will be created and saved to repository.
     */
    @Synchronized
    fun addGuild(
        guild: Guild,
        settings: BotGuildSettings?,
        invites: Collection<BotGuildInvite> = emptyList()
    ): BotGuild {
        val settingsToUse = if (settings == null) {
            val newSettings = BotGuildSettings()
            settingsRepository.set(guild, newSettings)
            newSettings
        } else {
            settings
        }

        val instance = BotGuild(this, guild, settingsToUse, invites)
        guilds[guild.idLong] = instance
        return instance
    }

    @Synchronized
    fun removeGuild(id: Long): BotGuild? {
        return guilds.remove(id)
    }

    @Synchronized
    fun getGuilds(): Set<BotGuild> {
        return HashSet(guilds.values)
    }

    @Synchronized
    fun stop() {
        log.info("Shutting down...")
        jda.shutdown()
        exitProcess(0)
    }

    fun createWarningMessage(warning: String): MessageEmbed {
        val footer = getMessageFooter()
        return EmbedBuilder()
            .setColor(Color.RED)
            .setTitle("Warning!")
            .setThumbnail(properties.getProperty("warning_icon_url"))
            .setDescription(warning)
            .setFooter(footer.first, footer.second)
            .build()
    }

    fun createInfoMessage(
        title: String,
        description: String? = null,
        fields: Collection<MessageEmbed.Field> = emptyList()
    ): MessageEmbed {
        val footer = getMessageFooter()
        val builder = EmbedBuilder()
            .setColor(Color.ORANGE)
            .setTitle(title)
            .setDescription(description)
            .setFooter(footer.first, footer.second)
        fields.forEach { builder.addField(it) }
        return builder.build()
    }

    private fun getMessageFooter(): Pair<String, String> {
        val text = "InviteRoles v${properties.getProperty("version")} by v${properties.getProperty("author.name")}"
        val icon = properties.getProperty("author.icon_url")
        return text to icon
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(InviteRoles::class.java)
    }
}

fun main(args: Array<String>) {
    // Parse args using Jopt-Simple - check it out, it's a cool little lib!
    val optionParser = OptionParser()
    val tokenSpec = optionParser.accepts("token").withRequiredArg().ofType(String::class.java).required()
    val dbPathSpec = optionParser.accepts("db").withRequiredArg().ofType(String::class.java).required()

    val optionSet = optionParser.parse(*args)
    val token = tokenSpec.value(optionSet)
    val dbPath = dbPathSpec.value(optionSet)

    val bot = InviteRoles(token, dbPath)
    bot.run()
}

/**
 * Replace last n characters with given character
 */
fun String.censorLast(count: Int = 3, char: Char = '•'): String {
    val chars = CharArray(length)
    for (i in 0 until length) {
        chars[i] = if (length - i <= count) char else this[i]
    }
    return String(chars)
}